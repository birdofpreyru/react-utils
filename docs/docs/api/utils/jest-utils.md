# Jest Utils
Collection of helpers useful in [Jest] tests.

:::caution
Prior to **v1.35.0** these utilities were exported as `JU` module of the main
package export, now they are exposed via a dedicated secondary export.

<details>
For example, to use [snapshot()] helper in a test prior to **v1.35.0** you would
do (and `JU` would be `null` when imported in non-`development` build, or outside
the server-side environment):
```tsx
import { JU } from '@dr.pogodin/react-utils';

test('Example', () => {
  JU!.snapshot(<div>Hello World</div>);
});
```
Starting with **v1.35.0** you must do:
```tsx
/** @jest-environment jsdom */

import { snapshot } from '@dr.pogodin/react-utils/jest';

test('Example', () => {
  snapshot(<div>Hello World</div>);
});
```
The change was done to avoid module resolution issues when using JSDom
environment for Jest tests.
</details>
:::

## E2eSsrEnv
Although it is not exported from this module for technical reasons, [E2eSsrEnv]
class implements Jest environment for end-to-end testing of SSR and client-side
execution of Webpack-built code with Jest, and thus should be considered as
a very important part of Jest-testing utilities provided by the library.

## Methods
- [act()](#act) &mdash; An alias for
  the [act()](https://reactjs.org/docs/test-utils.html#act) function from
  `react` (moved into `react` package since React v18.3).
- [getMockUuid()](#getmockuuid) - Generates a mock UUID (formats the given `seed`
  into an UUID-formatted string).
- [mockAxios()] &mdash; A helper which simplifies mocking out of [axios]
  (our selected HTTP(S) client library) in tests.
- [mockClientSide()] - Tricks **react-utils** into thinking the test is running
  within the client-side (browser) environment.
- [mockTimer()](#mocktimer) - Advances mock timers, and mock date by
  the specified time step.
- [mount()](#mount) - Mounts `scene` to the DOM, and returns the root scene
  element with **.destroy()** method attached.
- [snapshot()](#snapshot) - Does does a snapshot test of the give ReactJS
  component, and also returns JSON representation of is render.
- [unmockClientSide()](#unmockclientside) - Reverts the effect of previous
  [mockClientSide()] call.

:::caution Deprecated Methods
- [render()](#render) &mdash; Removed in **v1.34.0**. Migrate to
  [render()] provided by [@testing-library/react].
- [shallowRender()] &mdash; Removed in **v1.34.0**. Starting with React v19
  it is recommended to [avoid shallow rendering] in tests.
- [shallowSnapshot()](#shallowsnapshot) &mdash; Removed in **v1.34.0**.
  Starting with React v19 it is recommended to [avoid shallow rendering]
  in tests.
- [simulate](#simulate) &mdash; Removed in **v1.34.0**. Use instead
  [@testing-library/user-event], or the lower-level [fireEvent] from
  [@testing-library/dom].
:::

### act()
```tsx
import { act } from '@dr.pogodin/react-utils/jest';
```
This method is just an alias for
[act()](https://reactjs.org/docs/test-utils.html#act) function from
`react` (moved into `react` since v18.3).

### getMockUuid()
```tsx
import { getMockUuid } from '@dr.pogodin/react-utils/jest';

function getMockUuid(seed = 0): string;
```
Generates a mock UUID by determenistically transforming the given `seed` into
a UUID-formatted string.

**Arguments**
- `seed` &mdash; **number** &mdash; Optional. Defaults `0`.

**Returns**
- **string** - Mock UUID.

### mockAxios()
[mockAxios()]: #mockaxios
```ts
import { mockAxios } from '@dr.pogodin/react-utils/jest';

function mockAxios(handlers: AxiosRequestHandler[]): Axios;
```

[mockAxios()] is a helper which simplifies mocking out of [axios] (our selected
HTTP(S) client library) in tests. To use it, you want to create `__mock__/axios.ts`,
or `__mock__/axios/index.ts` file in the root of the host codebase:
```tsx
// Basic __mock__/axios.ts example.

import { type AxiosRequestHandlerT, mockAxios } from '@dr.pogodin/react-utils/jest';

const sampleHandler: AxiosRequestHandlerT = (config) {
  if (config.url === 'https://dr.pogodin.studio/docs/react-utils') {
    return { data: 'mock reponse' };
  }
};

export default mockAxios([
  sampleHandler,
]);
```

With such setup, every call to [axios] within Jest tests will be run through
each of provided mock request handlers; if any handler returns an object,
the corresponding [axios] call will be returned to that result (with any
omitted fields of [axios] response set to reasonable default values). If all
handlers return **null** or **undefined** results, the mock will fallback
to the real network call, and it will print a warning to the console, with
details of the corresponding request and response (which will help you to
mock that out).

For further details refer to [axios] documentation for
[request config](https://www.npmjs.com/package/axios#request-config)
(the argument of mock request handlers),
and [response schema](https://www.npmjs.com/package/axios#response-schema)
(the result of mock request handlers should be a
[partial](https://www.typescriptlang.org/docs/handbook/utility-types.html#partialtype)
of it).

### mockClientSide()
```tsx
import { mockClientSide } from '@dr.pogodin/react-utils/jest';

function mockClientSide();
```
Tricks **react-utils** library into thinking the test is running within
the client-side (browser) environment.

### mockTimer()
```tsx
import { mockTimer } from '@dr.pogodin/react-utils/jest';

async function mockTimer(time: number): Promise<void>;
```
Advances mock timers, and mock date by the specified time step.

**Arguments**

- `time` - **number** - The time step [ms].

**Returns**

- **Promise** - Resolves once all async operations triggered by the mock
  time advance have completed.

### mount()
```tsx
import { mount } from '@dr.pogodin/react-utils/jest';

function mount(scene: ReactNode): MountedSceneT;
```
Mounts given `scene` into DOM and returns the root scene element (DOM node) with
**.destroy()** and **.snapshot()** methods attached. The former of these methods
unmounts the scene from DOM; and the later snapshots the scene (_i.e._ is is
an alias for `expect(scene).toMatchSnapshot()`).

**Arguments**
- `scene` - **React.ReactNode** - The scene.

**Returns**
- **HTMLElement** - The root DOM element of the mounted scene with
  **.destroy()** and **.snapshot()** methods attached.

### render()
:::danger Deprecated
Removed in **v1.34.0**. Migrate to [render()] provided by [@testing-library/react].
:::

### shallowRender()
:::danger Deprecated
Removed in **v1.34.0**. Starting with React v19 it is recommended to
[avoid shallow rendering] in tests.
:::

### shallowSnapshot()
:::danger Deprecated
Removed in **v1.34.0**. Starting with React v19 it is recommended to
[avoid shallow rendering] in tests.
:::

### simulate()
:::danger Deprecated
Removed in **v1.34.0**. Use instead [@testing-library/user-event],
or the lower-level [fireEvent] from [@testing-library/dom].
:::

### snapshot()
[snapshot()]: #snapshot
```tsx
import { snapshot } from '@dr.pogodin/react-utils/jest';

// 1st overload (synchronous):
function snapshot(element: ReactElement): RenderResult;

// 2nd overload (asynchronous):
async function snapshot(
  element: ReactElement,
  options: SnapshotOptionsT,
): Promise<RenderResult>;
```
It does the snapshot test of given React component.

<details>
<summary>Example</summary>

```tsx
/** @jest-environment jsdom */

import { snapshot } from '@dr.pogodin/react-utils/jest';

test('A snapshot test', () => {
  snapshot(<div>Example</div>);
});
```
</details>

:::caution Beware
- It relies on [render()] function provided by [@testing-library/react],
  thus it renders provided component into DOM (**appends it to the current DOM
  content**), and snapshots HTML representation of the generated DOM node.

- It requires JSDom test environment, thus a Jest test relying on this method
  should start with the magic comment `/** @jest-environment jsdom */`, or it
  should provide a virtual DOM in an alternative way.

- Prior to **v1.34.0** this method relied on now deprecated
  [react-test-renderer](https://www.npmjs.com/package/react-test-renderer),
  and thus it was snapshotting and returning a JSON representation of the rendered
  component tree, without requiring or modifying the virtual DOM.
:::

**Arguments**
- `component` &mdash; **ReactElement** &mdash; React element to snapshot.
- `options` &mdash; [SnapshotOptionsT] | **undefined** &mdash; Optional settings.
  If provided, the function will be considered asynchronous, otherwise synchronous.

**Returns**
- [RenderResult](https://testing-library.com/docs/angular-testing-library/api/#renderresult) &mdash;
  render result from [@testing-library/react].

#### SnapshotOptionsT
[SnapshotOptionsT]: #snapshotoptionst
```ts
type SnapshotOptionsT = {
  await?: Promise<void>;
};
```
Optional settings for [snapshot()] method.
- `await` &mdash; **Promise&lt;void&gt;** | **undefined** &mdash; Optional.
  If provided, [snapshot()] will await for this promise prior to capturing
  the result of component render (it is awaited after the render is triggered).

### unmockClientSide()
```tsx
import { unmockClientSide } from '@dr.pogodin/react-utils/jest';

function unmockClientSide();
```
Reverts the effect of previous [mockClientSide()] call.

[@testing-library/dom]: https://testing-library.com/docs/dom-testing-library/intro
[@testing-library/react]: https://testing-library.com/docs/react-testing-library/intro/
[@testing-library/user-event]: https://testing-library.com/docs/user-event/intro
[avoid shallow rendering]: https://react.dev/blog/2024/04/25/react-19-upgrade-guide#removed-react-test-renderer-shallow
[axios]: https://www.npmjs.com/package/axios
[E2eSsrEnv]: /docs/api/classes/E2eSsrEnv
[fireEvent]: https://testing-library.com/docs/dom-testing-library/api-events
[Jest]: https://jestjs.io
[JU]: /docs/api/utils/jest-utils
[mockClientSide()]: #mockclientside
[render()]: https://testing-library.com/docs/react-testing-library/api#render
[shallowRender()]: #shallowrender
