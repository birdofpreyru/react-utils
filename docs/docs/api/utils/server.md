# server
```jsx
import { server } from '@dr.pogodin/react-utils';
```
The [server](/docs/api/utils/server) module provides constants and methods
facilitating implementation, setup and launch of a configurable ReactJS web
server.

:::info
Beside holding attached constants and functions documented below, `server`
export itself is also a function, which creates and launches a web-server for
ReactJS application. See [server()] documentation for further details.
:::

**Constants**
- [errors.CODES](#errorscodes) - A map from HTTP status code names to
  the corresponding numeric codes.
- [errors.ERRORS](#errorserrors) - A map from HTTP status code names to
  their pretty-printed names.
- [SCRIPT_LOCATIONS] - The map of valid `location` values in [Script] objects.

**Functions**
- [errors.assert()](#errorsassert) &mdash; Validates a value against the given
  validation schema compliant to the [Standard Schema Specs], _e.g._ [Zod]
  (recommended), or [Joi], _etc._
- [errors.fail()](#errorsfail) - Throws an error with the given message and
  HTTP status code.
- [errors.getErrorForCode()](#errorsgeterrorforcode) - Given HTTP status code
  returns the corresponding error text.

- [errors.newError()](#errorsnewerror) - Creates a new `Error` object with
  the given error message and attached HTTP status.
- [getDefaultCspSettings()](#getdefaultcspsettings) - Returns a deep copy of
  default [CSP] settings object used by **react-utils**.

:::caution Deprecated
- [errors.joi](#errorsjoi) &mdash; Removed in **v1.44.0**; an alias for [Joi]
  library, which provides methods for HTTP request validation.
:::

## Constants

### errors.CODES
[errors.CODES]: #errorscodes
```tsx
import { server } from '@dr.pogodin/react-utils';

const { CODES } = server.errors;

enum CODES {
  // Same as StatusCodes from 'http-status-codes'.
}
```
Enum of HTTP status codes &mdash; an alias of  **StatusCodes** enum from
the [http-status-codes] library.

**Example**
```jsx
import { server } from '@dr.pogodin/react-utils';

const { CODES } = server.errors;
console.log(CODES.BAD_REQUEST);

// Prints: 400
```

### errors.ERRORS
```jsx
server.errors.ERRORS: object
```
A map from HTTP status code names to their pretty-printed names. It is an
alias for **ReasonPhrases** object from [http-status-codes] library.

**Example**
```jsx
import { server } from '@dr.pogodin/react-utils';

const { ERRORS } = server.errors;
console.log(ERRORS.BAD_REQUEST);

// Prints: Bad Request
```

### SCRIPT_LOCATIONS
```jsx
server.SCRIPT_LOCATIONS: object
```
This object provides the map of valid [Script]'s `location` values:
- `BODY_OPEN` - Right after the opening `<body>` tag.
- `DEFAULT` - In the end of `<body>` tag, just before the main application
  bundle. This is the default location where scripts are injected when provided
  as strings rather than [Script] objects.
- `HEAD_OPEN` - Right after the opening `<head>` tag.

## Functions

### errors.assert()
```tsx
import { server } from '@dr.pogodin/react-utils';

const { assert } = server.errors;

async function assert<T extends StandardSchemaV1>(
  value: StandardSchemaV1.InferInput<T>,
  schema: T,
  message = '',
  statusCode = CODES.BAD_REQUEST, // 400
): Promise<StandardSchemaV1.InferOutput<T>>;
```
Validates the `value` against the given validation schema, compliant to the
[Standard Schema Specs], _e.g._ [Zod] (recommended), or [Joi], _etc._. Throws
an error with the given error `message` and `statusCode` if the validation fails,
otherwise resolves to the validated value.

**Arguments**
- `value` &mdash; **unknown** &mdash; The value to validate.
- `schema` &mdash; **StandardSchemaV1** &mdash; The validation schema, compliant
  to the [Standard Schema Specs].
- `message = ''` &mdash; **string** &mdash; Optional. Error message. If given,
  the error message from the validator will be appended to it, otherwise the
  error message from the validator is used as is.
- `statusCode = 400` &mdash; **number** &mdash; Optional. Error status code.
  Defaults `400` (bad request).

**Resolves** the validated value, with correctly inferred type.

### errors.fail()
```tsx
import { server } from '@dr.pogodin/react-utils';

const { CODES, fail } = server.errors;

function fail(
  message: string,
  statusCode: CODES = CODES.INTERNAL_SERVER_ERROR,
): Error;
```
Throws an error with the given `message` and HTTP `statusCode`.

**Arguments**
- `message` &mdash; **string** &mdash; Error message.
- `statusCode` - [errors.CODES] - Optional. HTTP status code. Defaults `500`
  (internal server error).

It never **returns**; however, its return type is declared as `Error` to allow
`throw fail(..)`, thus allowing the following TypeScript code:
```tsx
const value: number | undefined = someFunction();
if (value === undefined) throw fail('Some error');

// TypeScript allows this, because `throw fail()` above narrows down
// the `value` down to just `number`. 
const nextValue = value + 1;

// NOTE: According to TypeScript documentation it seems that `never` return
// type would be more appropriate for `fail()`; however, `never` return type
// does not result in the same value type narrowing in this example.
```

### errors.getErrorForCode()
```jsx
server.errors.getErrorForCode(code): string
```
Given HTTP status `code` returns the corresponding error text. It is an alias
for **getReasonPhrase()** function from [http-status-codes] library.

**Arguments & Result**
- `code` - **number** - HTTP status code.
- Returns **string** - Error message.

### errors.newError()
```jsx
server.errors.newError(message, statusCode = 500): Error
```
Creates a new `Error` object with given message, and HTTP `statusCode` attached
to the `result` as `result.status` field.

**Arguments & Result**
- `message` - **string** - Error message.
- `statusCode = 500` - **number** - Optional. HTTP status code. Defaults `500`
  (Internal Server Error).
- Returns **Error**.

### getDefaultCspSettings()
```jsx
server.getDefaultCspSettings(): object
```
Returns a deep copy of default [CSP] settings object used by **react-utils**,
with exception of `nonce-xxx` clause in `script-src` directive, which is added
dynamically for each request.

### errors.joi
:::danger Deprecated
Removed in **v1.44.0**. The original documentation below is kept for reference.

---

```tsx
server.errors.joi: object
```

An alias for [Joi] library, which provides methods useful for HTTP request
validation.

**Example**: [Joi] methods are used to create validation schema for the body of
a sample HTTP request.
```jsx
import { server } from '@dr.pogodin/react-utils';

const { joi } = server.errors;
const requestBodySchema = joi.object({
  sampleKey: joi.string().max(16).required(),
});
```

:::

[CSP]: https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP
[ExpressJS]: https://expressjs.com
[helmet]: https://github.com/helmetjs/helmet
[http-status-codes]: https://www.npmjs.com/package/http-status-codes
[Joi]: https://joi.dev/api
[Script]: /docs/api/functions/server#beforerender-script
[server]: /docs/api/utils/server
[SCRIPT_LOCATIONS]: #script_locations
[server()]: /docs/api/functions/server
[Standard Schema Specs]: https://standardschema.dev
[Webpack]: https://webpack.js.org
[Zod]: https://zod.dev
