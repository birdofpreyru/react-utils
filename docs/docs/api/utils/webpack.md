# webpack
```jsx
import { webpack } from '@dr.pogodin/react-utils';
```
Encapsulates Webpack-related utilities.

**Functions**
- [requireWeak()] - Requires a module at server side, and prevents it from being
  bundled by Webpack.
- [resolveWeak()](#resolveweak) - Resolves a module path with the help of
  Babel module resolver, thus the same way the resolution works for the normal
  `require()`.

## Functions

### requireWeak()
```tsx
import { webpack } from '@dr.pogodin/react-utils';

type RequireWeakOptionsT = {
  basePath?: string;
  throwOnError?: boolean;
};

webpack.requireWeak<Module extends NodeJS.Module>(
  modulePath: string,
  basePathOrOptions?: RequireWeakOptionsT | string,
): Module | null;
```
Implements server-side loading of the specified JS module in the way which
prevents it from being bundled into the client-side code by Webpack.

- **Server-side**: the `modulePath` argument is processed by Babel module
  resolver; if `basePath` is given, the resulting module path is additionally
  resolved from the base path using [path.resolve()]; the module is then loaded
  from the resulting path, and returned, similar to the regular `require()`, but
  invisible to Webpack, hence the module required this way is not bundled into
  the client side code. If the module could not be resolved, or the loading
  failed for other reason, [requireWeak()] returns **null**.

- **Client-side**: always returns **null**.

:::info Changes across `dr.pogodin/react-utils` versions:
- **v1.44.10**:
  Turns the second, optional argument into an object with `basePath` and
  `throwOnError` options. In past versions this optional argument was just
  `basePath` string, which still works but will be deprecated in future
  version.

- **v1.40.11**:
  - **TypeScript**: Added `Module` generic argument for typing of the result.
    You want to use it like:
    ```tsx
    import type M from 'some-module';

    const m = webpack.requireWeak<typeof M>('some-module');
    ```
- **v1.14.0**:
  - Safe to call at the client side, and always returns **null** there.
  - The `modulePath` argument is automatically processed by Babel module resovler.
  - `basePath` argument introduced.
:::

### resolveWeak()
```jsx
webpack.resolveWeak(modulePath): string
```
Resolves `modulePath` module path with the help of Babel module resolver,
thus the same way the resolution works for the normal `require()`.

**Arguments & Result**
- `modulePath` - **string** - Module path.
- Returns **string** - The resolved relative path to the module.

<!-- Reusable links. -->
[path.resolve()]: https://nodejs.org/dist/latest-v16.x/docs/api/path.html#pathresolvepaths
[requireWeak()]: #requireweak
