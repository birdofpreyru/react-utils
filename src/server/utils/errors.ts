/**
 * @category Utilities
 * @module server/errors
 * @desc
 * ```js
 * import { server } from '@dr.pogodin/react-utils';
 * const { errors } = server;
 * ```
 * Server-side helpers for error handling.
 */

import {
  StatusCodes as CODES,
  ReasonPhrases as ERRORS,
  getReasonPhrase as getErrorForCode,
} from 'http-status-codes';

import joi from 'joi';

/**
 * @static
 * @const CODES
 * @desc An alias for
 * [StatusCodes object from **http-status-codes** library](https://www.npmjs.com/package/http-status-codes).
 * It is a map between HTTP status code names and corresponding numeric codes.
 * @example
 * import { server } from '@dr.pogodin/react-utils';
 * const { CODES } = server.errors;
 * console.log(CODES.BAD_REQUEST); // Prints: 400
 */
export { CODES };

/**
 * @static
 * @const ERRORS
 * @desc An alias for
 * [ReasonPhrases object from **http-status-codes** library](https://www.npmjs.com/package/http-status-codes).
 * It is a map between HTTP status code names and their pretty-printed forms.
 * @example
 * import { server } from '@dr.pogodin/react-utils';
 * const { ERRORS } = server.errors;
 * console.log(ERRORS.BAD_REQUEST); // Prints: Bad Request
 */
export { ERRORS };

/**
 * @static
 * @func getErrorForCode
 * @desc An alias for
 * [getReasonPhrase() function from **http-status-codes** library](https://www.npmjs.com/package/http-status-codes).
 * Given an HTTP code it returns the corresponding error text.
 * @param {number} code HTTP code.
 * @return {string} HTTP error text.
 * @example
 * import { server } from '@dr.pogodin/react-utils';
 * console.log(server.errors.getErrorForCode(400)); // Prints: Bad Request
 */
export { getErrorForCode };

/**
 * @static
 * @const joi
 * @desc An alias for [Joi library](https://joi.dev/api/?v=17.4.0),
 * which provides tooling for HTTP request validation. You can use it in any
 * way you would use that library import.
 * @example
 * import { server } from '@dr.pogodin/react-utils';
 * const { joi } = server.errors;
 * const requestBodySchema = joi.object({
 *   sampleKey: joi.string().max(16).required(),
 * });
 */
export { joi };

// TODO: It could accept the status code as a constructor argument.
class ErrorWithStatus extends Error {
  status: number = CODES.INTERNAL_SERVER_ERROR;
}

/**
 * ```js
 * import { server } from '@dr.pogodin/react-utils';
 * const { newError } = server.errors;
 * ```
 * Creates a new `Error` object with given message, and HTTP status code
 * attached as `status` field.
 * @param {string} message Error message.
 * @param {number} [statusCode=500] HTTP status code. Defaults to 500 (Internal
 * Server Error).
 * @return {Error}
 */
export function newError(
  message: string,
  statusCode = CODES.INTERNAL_SERVER_ERROR,
) {
  const error = new ErrorWithStatus(message);
  error.status = statusCode;
  return error;
}

/**
 * Throws an error with given message and HTTP status code.
 */
export function fail(
  message: string,
  statusCode: CODES = CODES.INTERNAL_SERVER_ERROR,
): Error {
  throw newError(message, statusCode);
}

/**
 * ```js
 * import { server } from '@dr.pogodin/react-utils';
 * const { assert } = server.errors;
 * ```
 * Validates a value using given Joi schema, and throws an error with given
 * message and HTTP status code in case of the validation failure.
 * @param value
 * @param schema
 * @param [message] Error message.
 * @param [statusCode=500] HTTP status code. Defaults to 400 (Bad
 * Request).
 */
export function assert(
  value: unknown,
  schema: joi.AnySchema,
  message = '',
  statusCode = CODES.BAD_REQUEST,
) {
  const { error } = schema.validate(value, { abortEarly: false });
  if (error) {
    fail(message.concat(message ? '\n' : '', error.message), statusCode);
  }
}
