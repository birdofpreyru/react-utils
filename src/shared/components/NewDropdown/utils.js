/**
 * Auxiliary functions for <Dropdown> component.
 */

import { findIndex, findLastIndex } from 'lodash';

/**
 * Returns option name and value as two elements of array.
 * @param {object|string} option
 * @return {string[]}
 */
export function optionNameValue(option) {
  return typeof option === 'string'
    ? [option, option] : [option.name, option.value];
}

/**
 * Returns option value.
 * @param {object|string} option Option object or string.
 * @return {string}
 */
export function optionValue(option) {
  return typeof option === 'string' ? option : option.value;
}

/**
 * Finds the option by value.
 * @param {string} value Option value.
 * @param {array} options Options array.
 * @param {function} [filter] Optional. Filter function.
 * @return {number} Index of the matching option, or -1.
 */
function findOptionIndex(value, options, filter) {
  return options.findIndex(
    (item) => optionValue(item) === value && (!filter || filter(item)),
  );
}

/**
 * Finds the next option index.
 * @param {string} value Current value.
 * @param {array} options Option array.
 * @param {function} [filter] Optional. Filter function.
 * @return {number} Next option index, or -1.
 */
export function findNextOptionIndex(value, options, filter) {
  let index = findOptionIndex(value, options, filter);
  if (index >= 0 && ++index < options.length) {
    return filter ? findIndex(options, filter, index) : index;
  }
  return -1;
}

/**
 * Finds the previous option index.
 * @param {string} value Current value.
 * @param {array} options Option array.
 * @param {function} [filter] Optional. Filter function.
 * @return {number} Previous option index, or -1.
 */
export function findPrevOptionIndex(value, options, filter) {
  let index = findOptionIndex(value, options, filter) - 1;
  if (index >= 0 && filter) index = findLastIndex(options, filter, index);
  return index >= 0 ? index : -1;
}
