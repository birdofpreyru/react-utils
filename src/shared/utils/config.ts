/* global document */

import { parse } from 'cookie';

import clientGetInj from '../../client/getInj';

import { IS_CLIENT_SIDE } from './isomorphy/environment-check';
import { requireWeak } from './webpack';

// TODO: The internal type casting is somewhat messed up here,
// to be corrected later.
const config: Record<string, unknown> = (
  IS_CLIENT_SIDE ? clientGetInj().CONFIG : requireWeak('config')
) as (Record<string, unknown> | undefined) ?? ({} as Record<string, unknown>);

// The safeguard for "document" is necessary because in non-Node environments,
// like React Native, IS_CLIENT_SIDE is "true", however "document" and a bunch
// of other browser-world features are not available.
if (IS_CLIENT_SIDE && typeof document !== 'undefined') {
  config.CSRF = parse(document.cookie).csrfToken;
}

export default config;
