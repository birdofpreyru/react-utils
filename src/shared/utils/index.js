import themed, {
  COMPOSE,
  PRIORITY,
  ThemeProvider,
} from '@dr.pogodin/react-themes';

import config from './config';
import * as isomorphy from './isomorphy';
import time from './time';
import * as url from './url';
import * as webpack from './webpack';

themed.COMPOSE = COMPOSE;
themed.PRIORITY = PRIORITY;

let JU = null; // eslint-disable-line import/no-mutable-exports
if (isomorphy.IS_SERVER_SIDE
  && (process.env.NODE_CONFIG_ENV || process.env.NODE_ENV) !== 'production'
) {
  /* eslint-disable global-require */
  const path = webpack.requireWeak('path');
  JU = webpack.requireWeak(path.resolve(__dirname, './jest'));
  /* eslint-enable global-require */
}

/**
 * Creates a new async barrier: a Promise instance with its resolve method
 * attached as .resolve field.
 * @return {Promise<>}
 */
function newBarrier() {
  let resolve;
  const promise = new Promise((res) => { resolve = res; });
  promise.resolve = resolve;
  return promise;
}

/**
 * Attempts to execute provided async "action" up to "maxRetries" time with
 * the given "interval". If any of attempts is successful it returns its result,
 * otherwise if all attempts fail it throws the error of last attempt.
 * @param {function} action
 * @param {number} [maxRetries=5] Optional. Maximum number of retries. Defaults
 *  to 5 attempts.
 * @param {number} [interval=1000] Optional. Interval between retries [ms].
 *  Defaults to 1 second.
 * @return {Promise} Resolves to the result of successful operation, or
 *  rejects with the error from the latst failed attempt.
 */
async function withRetries(action, maxRetries = 5, interval = 1000) {
  /* eslint-disable no-await-in-loop */
  for (let n = 1; ; ++n) {
    try {
      return await action();
    } catch (error) {
      if (n < maxRetries) await time.timer(interval);
      else throw error;
    }
  }
  /* eslint-enable no-await-in-loop */
}

export {
  config,
  isomorphy,
  JU,
  newBarrier,
  themed,
  ThemeProvider,
  time,
  url,
  webpack,
  withRetries,
};
