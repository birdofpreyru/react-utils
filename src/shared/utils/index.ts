import themedImpl, {
  COMPOSE,
  PRIORITY,
  type Theme,
  ThemeProvider,
} from '@dr.pogodin/react-themes';

import config from './config';
import * as isomorphy from './isomorphy';
import time from './time';
import * as webpack from './webpack';

export {
  type Listener,
  Barrier,
  Emitter,
  Semaphore,
  withRetries,
} from '@dr.pogodin/js-utils';

export { getSsrContext } from './globalState';
export { default as splitComponent } from './splitComponent';

type ThemedT = typeof themedImpl & {
  COMPOSE: typeof COMPOSE;
  PRIORITY: typeof PRIORITY;
};

const themed: ThemedT = themedImpl as ThemedT;

themed.COMPOSE = COMPOSE;
themed.PRIORITY = PRIORITY;

export {
  type Theme,
  config,
  isomorphy,
  themed,
  ThemeProvider,
  time,
  webpack,
};
