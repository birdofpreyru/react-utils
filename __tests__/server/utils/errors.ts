import { fail } from 'server/utils/errors';

test('fail()', () => {
  let error: unknown;
  let never: unknown;
  try {
    never = fail('Mock error', 418);
  } catch (e: unknown) {
    error = e;
  }
  expect(error).toMatchSnapshot();
  expect((error as { status: number }).status).toBe(418);
  expect(never).toBeUndefined();
});
