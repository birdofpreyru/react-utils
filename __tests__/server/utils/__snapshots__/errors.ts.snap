// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`fail() 1`] = `[Error: Mock error]`;
